feat: adds REINFORCE algorithm (#357)
NVIDIA/NeMo-Aligner: docs/user-guide/reinforce.rst

Reviewer: For line 33, CHECKPOINT_NEMO_FILE should be changed to RM_NEMO_FILE

Author: Thanks, fixed.

#. The Initial Policy Network (also known as the Reference Model): We use this model to compute a KL Divergence penalty term that ensures that the PPO Actor does not diverge too much from the Initial Policy. This way, we prevent the REINFORCE Actor from overfitting to the rewards given by the RM, and ensure it does not forget the knowledge it acquired during pretraining and SFT. This model should be the one used to initialize the REINFORCE Actor Network.

The next section discusses how to launch each of these two jobs.

Launching the Reward Model and Critic Server
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

To launch the server:

.. code-block:: bash 

   #!/bin/bash
   CHECKPOINT_NEMO_FILE="/path/to/trained_rm.nemo"
   GPFS="/path/to/nemo-aligner-repo"

   RESULTS_DIR="critic_results_dir"

   export PYTHONPATH="${GPFS}:${PYTHONPATH}" \
   && export HYDRA_FULL_ERROR=1 \
   && python -u examples/nlp/gpt/serve_reward_model.py \
      trainer.num_nodes=1 \
      trainer.devices=8 \
      ++model.tensor_model_parallel_size=4 \
      rm_model_file=${RM_NEMO_FILE}

> The above example launches the reward model server on 8 GPUs and 1 node. Please make sure to change trainer.devices, trainer.num_nodes depending on your model size and scale. Aligner will work on any scale. Also, make sure to tune the trainer.reinforce.inference_micro_batch_size argument. This argument sets the size of the batch the REINFORCE actor is allowed to send to the reward per DP rank.

Reviewer 2: ```suggestion
The above example launches the reward model server on eight GPUs and one node. Make sure to change trainer.devices, trainer.num_nodes depending on your model size and scale. Aligner will work on any scale. Also, make sure to tune the trainer.reinforce.inference_micro_batch_size argument. This argument sets the size of the batch the REINFORCE actor is allowed to send to the reward per DP rank.
```

Author: Fixed.

      ++model.tensor_model_parallel_size=4 \
      ++model.reinforce.num_rollout_samples=${NUM_ROLLOUTS} \
      model.global_batch_size=${ACTOR_GBS} \
      model.micro_batch_size=1 \
      model.optim.lr=\"\\\$\{multiply:${ACTOR_LR},1.001\}\" \
      model.optim.sched.warmup_steps=0 \
      model.optim.sched.constant_steps=312 \
      model.optim.sched.min_lr=${ACTOR_LR} \
      model.optim.weight_decay=0.01 \
      model.reinforce.rollout_micro_batch_size=16 \
      model.reinforce.forward_micro_batch_size=16 \
      model.reinforce.val_rollout_micro_batch_size=8 \
      model.data.data_impl=jsonl \
      remote_rm.reward_model.ip=${host_reward} \
      remote_rm.reward_model.port=${REWARD_PORT} \
      ++model.reinforce.length_params.max_length=2048 \
      trainer.reinforce.initial_policy_kl_penalty="${KL}" \
      ++model.optim.bucket_cap_mb=200 \
      ++model.dist_ckpt_format=zarr \
      ++model.optim.overlap_grad_sync=False \
      ++model.optim.contiguous_grad_buffer=True \
      ++model.enable_nge=True \
      trainer.reinforce.batch_iterator.use_flask=${USE_FLASK} \
      trainer.reinforce.rollout_batch_seq_length=4096

> The above command launches the initial and actor server on 1 node with 8 GPUs.

Reviewer 2: ```suggestion
The above command launches the initial and actor server on one node with eight GPUs.
```

Author: Changed.

Launching Both Servers for REINFORCE training
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

> You can use slurm to launch the 2 jobs and get them to coordinate together in a full REINFORCE job via the following:

Reviewer 2: ```suggestion
You can use slurm to launch the two jobs and get them to coordinate together in a full REINFORCE job through the following:
```

Author: Fixed.

      model.global_batch_size=${ACTOR_GBS} \
      model.micro_batch_size=1 \
      model.optim.lr=\"\\\$\{multiply:${ACTOR_LR},1.001\}\" \
      model.optim.sched.warmup_steps=0 \
      model.optim.sched.constant_steps=312 \
      model.optim.sched.min_lr=${ACTOR_LR} \
      model.optim.weight_decay=0.01 \
      model.reinforce.rollout_micro_batch_size=16 \
      model.reinforce.forward_micro_batch_size=16 \
      model.reinforce.val_rollout_micro_batch_size=8 \
      model.data.data_impl=jsonl \
      remote_rm.reward_model.ip=${host_reward} \
      remote_rm.reward_model.port=${REWARD_PORT} \
      ++model.reinforce.length_params.max_length=2048 \
      trainer.reinforce.initial_policy_kl_penalty="${KL}" \
      ++model.optim.bucket_cap_mb=200 \
      ++model.dist_ckpt_format=zarr \
      ++model.optim.overlap_grad_sync=False \
      ++model.optim.contiguous_grad_buffer=True \
      ++model.enable_nge=True \
      trainer.reinforce.batch_iterator.use_flask=${USE_FLASK} \
      trainer.reinforce.rollout_batch_seq_length=4096
   EOF

   srun --het-group=1 -o $PPO_OUTFILE -e $PPO_ERRFILE --container-image=${CONTAINER} $MOUNTS bash -c "${cmd_reinforce}" &

   wait

> The above script runs the reward model server on 1 node and the actor on 1 node.

Reviewer 2: ```suggestion
The above script runs the reward model server on one node and the actor on one node.
```

Author: Fixed.

> It is important to launch all jobs with ``&`` after the srun command, to ensure they do not block each other. 

Reviewer 2: ```suggestion
It is important to launch all jobs with ``&`` after the srun command to ensure they do not block each other. 
```

Author: Fixed.

> .. note::

Reviewer 2: ```suggestion
.. Note::
```

Author: Fixed.

   Make sure to change the reward model arg ``trainer.reinforce.inference_micro_batch_size`` such that ``trainer.reinforce.inference_micro_batch_size * DP size <= model.reinforce.rollout_micro_batch_size``.

REINFORCE Results
%%%%%%%%%%%%%%%%%%%%%%%%%%

> Once you've completed reinforce training, you can serve your model using the `megatron_gpt_eval.py <https://github.com/NVIDIA/NeMo/blob/8cd5f1c8e7d4fed9f4f946028cd02047c5d2296f/examples/nlp/language_modeling/megatron_gpt_eval.py#L4>`__ script from the NeMo codebase to run more rigorous evaluation of your trained model.

Reviewer 2: ```suggestion
After you've completed reinforce training, you can serve your model using the `megatron_gpt_eval.py <https://github.com/NVIDIA/NeMo/blob/8cd5f1c8e7d4fed9f4f946028cd02047c5d2296f/examples/nlp/language_modeling/megatron_gpt_eval.py#L4>`__ script from the NeMo codebase to run more rigorous evaluation of your trained model.
```

      trainer.num_nodes=1 \
      trainer.devices=8 \
      ++model.tensor_model_parallel_size=4 \
      rm_model_file=${RM_NEMO_FILE}

The above example launches the reward model server on 8 GPUs and 1 node. Please make sure to change trainer.devices, trainer.num_nodes depending on your model size and scale. Aligner will work on any scale. Also, make sure to tune the trainer.reinforce.inference_micro_batch_size argument. This argument sets the size of the batch the REINFORCE actor is allowed to send to the reward per DP rank.

Launch the Initial Policy and REINFORCE Actor Training
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

The REINFORCE Actor training job contains the master controller that makes the HTTP calls to all servers when needed. To launch the REINFORCE Actor and Initial Policy server:

.. code-block:: bash 

   GPFS="/path/to/nemo-aligner-repo"
   TRAIN_DATA_PATH="/path/to/train_prompts.jsonl"
   VALID_DATA_PATH="/path/to/test_prompts.jsonl"

   PRETRAINED_ACTOR_NEMO_FILE="/path/to/sft_checkpoint.nemo"
   RESULTS_DIR="/path/to/actor_results_dir"

   USE_FLASK=False
   ACTOR_LR=1e-6
   KL=0.01
   NUM_ROLLOUTS=32
   ACTOR_GBS=32
   REWARD_PORT=5555

>    host_reward="$(scontrol show hostnames=$SLURM_JOB_NODELIST_HET_GROUP_0 | head -n1)"

Author: Changed reward_host

      model.data.data_impl=jsonl \
      remote_rm.reward_model.ip=${host_reward} \
      remote_rm.reward_model.port=${REWARD_PORT} \
      ++model.reinforce.length_params.max_length=2048 \
      trainer.reinforce.initial_policy_kl_penalty="${KL}" \
      ++model.optim.bucket_cap_mb=200 \
      ++model.dist_ckpt_format=zarr \
      ++model.optim.overlap_grad_sync=False \
      ++model.optim.contiguous_grad_buffer=True \
      ++model.enable_nge=True \
      trainer.reinforce.batch_iterator.use_flask=${USE_FLASK} \
      trainer.reinforce.rollout_batch_seq_length=4096

The above command launches the initial and actor server on 1 node with 8 GPUs.

Launching Both Servers for REINFORCE training
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

You can use slurm to launch the 2 jobs and get them to coordinate together in a full REINFORCE job via the following:

.. code-block:: bash 

   #!/bin/bash
   #SBATCH -N 1 --ntasks-per-node 8 -A <<ACCOUNT>> -p <<PARTITION>> --job-name <<JOBNAME>> -t 4:00:00 --exclusive
   #SBATCH hetjob
   #SBATCH -N 1 --ntasks-per-node 8 -A <<ACCOUNT>> -p <<PARTITION>> --job-name <<JOBNAME>> -t 4:00:00 --exclusive

>    NAME="2p_reinforce"

Author: Removed it from the name.
